src/backend/tcop/postgres.c

```diff
@@ -1592,7 +1592,8 @@ send_guc_to_QE(List *guc_list, bool is_restore)
 			/* if some guc can not restore successful
 			 * we can not keep alive gang anymore.
 			 */
-			DisconnectAndDestroyAllGangs(false);
+			DisconnectAndDestroyAllGangs(true);
+			CheckForResetSession();
```

Comment:
Why do we need to call the `CheckForResetSession` function here? Without this call, the new test passes successfully.

Reply:
`CheckForResetSessions` expects that it is called immediately (or at least soon) after `DisconnectAndDestroyAllGangs`, because it expects that no gangs are present. If it is called at some point later when a new gang is allocated some [other tests](https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1901729) fail (and in general behavior is really unstable)

Reply:
Why then isn't this `CheckForResetSession` function always called immediately after the `DisconnectAndDestroyAllGangs` function elsewhere?

Reply:
Because it is called in [main](https://github.com/arenadata/gpdb/blob/2ce272a173b3353bcb45723ed8b9aa887025f18d/src/backend/tcop/postgres.c#L5273) before every command. However, GUC sync happens after `CheckForResetSession`, but before executing the actual command, so the command can again create a new session. In all other cases `DisconnectAndDestroyAllGangs` is called before new command starts.